### ISSUE TITLE
Bizarre navigation at Sundance dr & Sentinel dr in Fremont, CA, USA

### ISSUE THREAD

**Issue by CommanderPrS** [osmandapp/Osmand#1743]
OsmAnd version: 2.1.1 (latest from F-Driod)

![osmand](https://cloud.githubusercontent.com/assets/14830158/10091301/5120c964-62ee-11e5-88e4-dc26d734df1c.png)

California map version: 2015/09/06 (latest)

![maps](https://cloud.githubusercontent.com/assets/14830158/10091316/8480857e-62ee-11e5-85c9-ffe6a97e1980.png)

It seems that the intersection of Sundance dr and Sentinel dr in Fremont, CA, USA is problematic for navigation (I am using offline maps and offline navigation). OsmAnd appears to refuse to make a left turn from Sundance to Sentinel (or right turn in reverse) in some situations.

__Case 1: Going to Sentinel dr from anywhere south-west from it__

Let's pick a random destination, for example this one. The precise location is not critical, it happens from pretty much anywhere.

![source](https://cloud.githubusercontent.com/assets/14830158/10091368/555a18ea-62ef-11e5-979d-cca2f1b32e57.png)

And then choose to go somewhere on Sentinel dr a little bit to the north from this intersection. Notice that the suggested route refuses to take Sundance dr even though it is obviously shorter:

![destination](https://cloud.githubusercontent.com/assets/14830158/10091364/3e0b8142-62ef-11e5-98cc-d6d0db20af23.png)

__Case 2: Force a via point on Sundance dr__

If I try to force the route to go through Sundance dr by adding an explicit via point, the navigation actually does not reach the destination - it stops at the intersection!

![viapoint](https://cloud.githubusercontent.com/assets/14830158/10091414/df26cfb4-62ef-11e5-8582-3c0cf6f15e8f.png)

__Case 3: Going further up on Sentinel dr__

My thinking at this point - maybe the vector information in the map is wrong and somehow the intersection does not connect the two streets. This is not true. If I move the destination just a little bit further north on Sentinel dr, suddenly the route is picked correctly.

![destinationup](https://cloud.githubusercontent.com/assets/14830158/10091428/1a51dd0e-62f0-11e5-9137-e5feec23d87b.png)

__Case 4: Going in reverse from Sentinel dr__

The reverse routing also does not like this intersection. If I choose to from the destination in case 1 to roughly the same source in case 1 (see the next screenshot), I end up with a route that avoids Sundance dr.

The reverse destination:

![reversedestination](https://cloud.githubusercontent.com/assets/14830158/10091466/a497aa48-62f0-11e5-8212-0ba4a0a56669.png)

The resulting route:

![reversesource](https://cloud.githubusercontent.com/assets/14830158/10091470/ab3a6ba6-62f0-11e5-99d3-69e8ff924a84.png)

__Case 5: Force a via point on Sundance dr in the reverse direction__

If I add an explicit via point on Sundance dr, then the navigation picks up the correct route through Sundance dr.

![reverseviapoint](https://cloud.githubusercontent.com/assets/14830158/10091504/12ac4656-62f1-11e5-9c3b-d2584f7bdb55.png)

__Case 6: Going from further up on Sentinel dr__

The interesting part is that if I try to move the source up north on Sentinel dr to the point where the forward navigation worked correctly (case 3 above), the reverse navigation still does NOT pick Sundance dr and I have to force the route to go through it using a via point.

![reversesourceup](https://cloud.githubusercontent.com/assets/14830158/10091522/52d71b20-62f1-11e5-8c1d-7cd2999e0e09.png)

__Case 7: Going from even further up on Sentinel dr__

If I move the source point further up north on Sentinel dr, then the navigation picks up Sundance dr without any issues and I do not need to use a via point to force the correct route.

![reversesourcefurtherup](https://cloud.githubusercontent.com/assets/14830158/10091541/93f8be42-62f1-11e5-9399-055c6a11891a.png)

__Summary__

Bottom line - it looks like the navigation system does not like this intersection. Even though the map does allow turning through that intersection, the navigation logic avoids it depending on the exact location of the source/destination on Sentinel dr.

As a validation, I tried going between the same points on openstreetmaps.net using OSRM navigation and it found the correct route (through Sundance dr) in both directions without any issues.

**Comment by Gizmot** [comment]
Looked at OSM and didnot see anything obvious.
Is it possible that the stream underneath cause some routing bug?

**Comment by CommanderPrS** [comment]
> Is it possible that the stream underneath cause some routing bug?

Hard to say. As far as I understand, if the stream were interfering with the routing, then it would be permanent. For example, if the stream were to cut Sundance dr into two disconnected segments, then it would be impossible to force the route to go through Sundance dr - yet I was able to do so using via points both in the forward and in the reverse direction.

If I had to guess, it seems that this bizarre behavior could be explained by the routing algorithm refusing to pick a route that would need to make a turn and immediately stop. This is why it was picking the round-about way in Case1 above.

To test that, I tried routing from the same source south-west to a point on Sentinel dr a little bit further north. And yes, the route is picked correctly though Sundance dr without any via points. Apparently, the stream is not a problem here:

![destinationupnoviapoint](https://cloud.githubusercontent.com/assets/14830158/10110061/b2f04052-637f-11e5-8ccb-851c9e300489.png)

Going in reverse (from Sentinel dr), the routing algorithm seems to dislike making a turn too quickly after starting the route. This would explain Case 7 above. The only difference between routing to/from Sentinel seems to be the tolerance for the short turn - the tolerance for a turn at the end seems to be smaller (shorter turns allowed) than the tolerance at the beginning at the route. This would explain why for the reverse route I had to move the starting point further up Sentinel dr (compare Case 6 and Case 7 above).

BTW, the reason I am talking about short turns is because I am picking the point on Sentinel dr very close to the intersection - 50 meters or so. Picking a point additional 50 meters (100 meters total) away from the intersection makes the "to" route happy. Picking a point even further 50 meters up (150 meters total) from the intersection makes the "from" route happy.

**Comment by utack** [comment]
Possible duplicate of either #1195 or #1468 ?

**Comment by vshcherb** [comment]
May be the issue was with map data, not reproducible any more. https://www.dropbox.com/s/e9gojq0z902qrcp/Screenshot%202015-11-07%2013.49.19.png?dl=0

**Comment by CommanderPrS** [comment]
NOTE: The system does not allow me to reopen this bug. I had to open another issue to do that. Ideally, this bug should be reopen and issue #2063 closed to avoid duplicating bugs.


> May be the issue was with map data, not reproducible any more.

I waited for 2 months to make sure any new data you had gets propagated to the public map repository. After that I updated the maps. Here is the screenshot:

![screenshot_2015-12-15-15-18-10](https://cloud.githubusercontent.com/assets/14830158/11827442/25499f58-a341-11e5-89fc-3f43d6c3fdba.png)

The issue is still present. OsmAnd is still not finding the right route. I doubt this is caused by the map data. If it were, simply moving the start (or end) point of the route up-down Sentinel dr would not change the route as drastically as it happens. Please take a look at the screenshots for the experiments I did when filing this bug - I was specifically trying to diagnose whether this is a map problem or a routing problem.

In fact, I suspect that even in your experiment if you move the point on Sentinel dr as shown in the diagram below you would get a route that would NOT select Sundance dr.

![screenshot 2015-11-07 13 49 19](https://cloud.githubusercontent.com/assets/14830158/11827538/b6945584-a341-11e5-8d64-232e7edee99b.png)

**Comment by CommanderPrS** [comment]
BTW, I was testing with the starting end of the route being a bit further on Mission blvd. Please see the screenshot in Case 1 above. Maybe your start point on Klamath street is too close to trigger this bug.

With my starting point, OsmAnd selects an obviously wrong route as shown in this screenshot. And yes, I can force the correct route by adding a way point on Sundance dr.

![screenshot_2015-12-15-15-42-28](https://cloud.githubusercontent.com/assets/14830158/11827731/5219c236-a343-11e5-9992-750a4f0a3218.png)

**Comment by vshcherb** [comment]
Main issue is fixed

**Comment by CommanderPrS** [comment]
> Main issue is fixed

Today I installed 2.4.4 and verified that the route was calculated correctly. Thus, I confirm that this issue is fixed.
